test/partition-healing-tests.js

```diff
@@ -0,0 +1,354 @@
+var safeJSONParse = require('./util').safeParse;
+var test2 = require('./test-util').test2;
+
+// First stage of healing algorithm: forced reincarnation.
+//
+// Actors:
+// ● a1, sut: partition A.
+// ● b1, b2: partition B.
+//
+// 1. Create two partitions A and B by marking b1 and b2 failed in sut.
+//    a1 and sut are failed according to B.
+// 2. Trigger partition healing.
+// 3. Verify sut tries to reincarnate both sides:
+// 3.1. sut declares b1 and b2 as suspects to B.
+// 3.2. sut declares a1 as suspect to A.
+test2('reincarnating partitions A and B', [3], 20000,
+        prepareCluster(function(t, tc, n) {
+
+            return [
+                dsl.changeStatus(t, tc, 1, 1, 'faulty', 0),
+                dsl.changeStatus(t, tc, 2, 2, 'faulty', 0),
+                dsl.waitForPingResponse(t, tc, 1),
+                dsl.waitForPingResponse(t, tc, 2),
+                dsl.assertStats(t, tc, 2, 0, 2),
+
+                // According to B, a1 is faulty.
+                makeAFaultyInB(t, tc),
+
+                // Trigger partition heal
+                dsl.callEndpoint(t, tc, "/admin/healpartition/disco", {},
+                        function(eventBody) { validateHealRequest(t, tc, eventBody) }),
+                removeHealPartitionDiscoResponse(t, tc),
+
+                // Verify join request came in to b1 or b2
+                dsl.waitForJoins(t, tc, 1),
+
+                // Verify b1 and b2 are declared suspects, but 'a' is not disseminated as alive.
+                verifySuspects(t, tc, [1, 2], [0]),
+
+                // SUT tries to reincarnate partition A -- send a ping to a1.
+                // Also verify it didn't have information about b1: that means
+                // sut didn't try to merge partitions yet.
+                verifySuspects(t, tc, [0], [1, 2])
+            ]
+        }
+        )
+     );
+
+// Assign names to nodes and update membership of B from test coordinator.
+function makeAFaultyInB(t, tc) {
+    return function makeAFaultyInB(list, cb) {
+        tc.fakeNodes[1].status = 'alive';
+        tc.fakeNodes[2].status = 'alive';
+
+        var B = tc.getMembership();
+        // According to B, a1 is faulty
+        _.find(B, {port: tc.fakeNodes[0].port}).status = 'faulty';
+        _.find(B, {port: tc.fakeNodes[0].port}).incarnationNumber = 1337;
+
+        tc.fakeNodes[1].membership = _.cloneDeep(B);
+        tc.fakeNodes[2].membership = _.cloneDeep(B);
+
+        cb(list);
+    }
+}
+
+function removeHealPartitionDiscoResponse(t, tc) {
+    return function removeHealPartitionDiscoResponse(list, cb) {
+        var d = _.find(list, { type: events.Types.AdminHealPartitionDisco });
+        if (d === undefined) {
+            cb(null);
+        } else {
+            list.splice(list.indexOf(d), 1);
+            cb(list);
+        }
+    }
+}
+
+// Verify 'nodes' are treated as suspects by sut. Also, verify sut is not
+// sending 'nodes' information that 'nodes_in_other_partition' are alive.
```

Comment:
Can we add a comment along the lines of: verifySuspect waits for a ping from the sut that contains the following changes: nodes from nodes are declared suspect, and no nodes from nodes_in_other_partition are declared alive.